### ISSUE TITLE
`sycl::buffer::set_final_data` is tested with `std::shared_ptr` which is not required by the spec

### ISSUE THREAD

**Issue by nilsfriess** [KhronosGroup/SYCL-CTS#877]
In the spec, the templated type `Destination` of the parameter `finalData` that is passed to `sycl::buffer::set_final_data` is described as:
> Destination can be either an output iterator or a std::weak_ptr<T>.

In the buffer tests (specifically in [this file](https://github.com/KhronosGroup/SYCL-CTS/blob/SYCL-2020/tests/buffer/buffer_storage_common.h), lines 66 and 69), `set_final_data` is also tested with `Destination == std::weak_ptr<T[]>` and `Destination == std::shared_ptr<T[]>`, and I don't think that these types are required as per the spec.

So I think either the spec has to be extended or these test cases should be removed from the CTS (and a test case for `Destination == std::weak_ptr<T>` should be added).

**Comment by gmlueck** [comment]
I agree that the spec is not clear and that there is a disconnect with the test suite.  Looking at the DPC++ implementation, it seems that we allow `Destination` to be any type that is convertible to `std::weak_ptr`, though the spec doesn't say that.  This is why DPC++ accepts `std::shared_ptr`.

The spec is also not clear about what type of `std::weak_ptr` is allowed for `Destination`.  Must it be `std::weak_ptr<T>`, or can it be `std::weak_ptr<T[]>`?  What about `std::weak_ptr<U>`, where `T` is convertible to `U`?

We might also want to tighten up the specification to say that the output iterator must allow writing objects of type `T`.

**Comment by keryell** [comment]
I think this is because at some point in the SYCL 1.2 specification there was some special handling for `std::shared_ptr` but also `std::shared_ptr` in some other places. Then I realized that they were redundant because of some implicit conversions.

**Comment by gmlueck** [comment]
Implicit conversions happen only after template selection, though.  The problem is that `set_final_data` is defined like:

> ```
> template <typename Destination = std::nullptr_t>
> void set_final_data(Destination finalData = nullptr);
> ```
>
> _Constraints:_ Available only when `Destination` is an output iterator or is `std::weak_ptr<T>`.

If the user calls `set_final_data` with `std::shared_ptr<T>`, then the template doesn't add any candidates to the overload set because _Destination_ doesn't satisfy the constraints.  If we want to allow `std::shared_ptr<T>` as an argument type, we need to add it somehow to the constraints.
